api/src/auth/endpoint_access_util.py

```diff
@@ -113,3 +115,16 @@ def get_log_info_for_resource(resource: Organization | Application | Agency | No
         log_info["agency_code"] = resource.agency_code
 
     return log_info
+
+
+def check_user_access(
+    db_session: db.Session,
+    user: User,
+    allowed_privileges: set[Privilege],
+    resource: Organization | Application | Agency | None,
+) -> None:
+    """Wrapper function to preload all roles and privileges for the given user in a single query,
+    then checks access using the in-memory data."""
+    user_with_role = get_roles_and_privileges(db_session, user.user_id)
+    if user_with_role and not can_access(user_with_role, allowed_privileges, resource):
```

Comment:
It shouldn't happen, but can we instead make this error if no user was found? With the way our auth works, the user needs to exist to even be authenticated in the first place (minus that one weird endpoint with the alt auth, but it shouldn't call this for that scenario).

```suggestion
    if not user_with_role or not can_access(user_with_role, allowed_privileges, resource):
```